### PR TITLE
[luci-interpreter] Support Integer Sub

### PR DESCRIPTION
**Pull request by BalyshevArtem**
This commit enables integer Sub kernels.

Related issue: #8155

ONE-DCO-1.0-Signed-off-by: Artem Balyshev a.balyshev@partner.samsung.com

### CODE REVIEW

File: compiler/luci-interpreter/src/kernels/Sub.test.cpp [hunk 1]
```diff
@@ -162,6 +162,52 @@ TEST_F(SubTest, Float)
...
+     0, 2, 4, 4, 0, 0, 3, 0, 10, 0, 6, 0, 3, 0, 10, 2, 6, 0},
+    {0, 1, 4, 1, 3, 0, 0, 2, 10, 0, 6, 0},
+    {0, 0, 0, 1, 2, 5, 0, 0, 0, 0, 4, 3, 0, 0, 3, 0, 7, 0,
+     2, 4, 0, 2, 0, 0, 8, 0, 6, 0, 1, 0, 8, 2, 6, 0, 1, 0},
+    {0, 0, 0, 0, 7, 0, 2, 4, 6, 0, 1, 0}};
+  std::vector<int32_t> input1_data{-1, 2, 1, 0, 4, -5, 1, 3, 7, -1, 7, 1};
+  std::vector<int32_t> input2_data{4, 1, -3, -1, 1, 6};
+  for (size_t i = 0; i < test_shapes.size(); ++i)
+  {
+    Tensor input1_tensor = makeInputTensor<DataType::S32>(base_shape, input1_data, memory_manager);
```
Comment (binarman) [inline compiler/luci-interpreter/src/kernels/Sub.test.cpp:181]:
```suggestion
    Tensor input1_tensor = makeInputTensor<DType>(base_shape, input1_data, memory_manager);
```

File: compiler/luci-interpreter/src/kernels/Sub.test.cpp [hunk 2]
```diff
@@ -162,6 +162,52 @@ TEST_F(SubTest, Float)
...
+    Tensor input2_tensor =
+      makeInputTensor<DataType::S32>(test_shapes[i], input2_data, memory_manager);
```
Comment (binarman) [inline compiler/luci-interpreter/src/kernels/Sub.test.cpp:183]:
```suggestion
      makeInputTensor<DType>(test_shapes[i], input2_data, memory_manager);
```

File: compiler/luci-interpreter/src/kernels/Sub.test.cpp [hunk 3]
```diff
@@ -162,6 +162,52 @@ TEST_F(SubTest, Float)
...
+    Tensor output_tensor = makeOutputTensor(DataType::S32);
```
Comment (binarman) [inline compiler/luci-interpreter/src/kernels/Sub.test.cpp:184]:
```suggestion
    Tensor output_tensor = makeOutputTensor(DType);
```

File: compiler/luci-interpreter/src/kernels/Sub.test.cpp [hunk 4]
```diff
@@ -162,6 +162,52 @@ TEST_F(SubTest, Float)
...
+  std::vector<Shape> test_shapes{{1, 1, 3, 2}, {1, 3, 1, 2}, {2, 1, 3, 1}, {2, 3, 1, 1}};
+  std::vector<std::vector<int32_t>> test_outputs = {
+    {0, 1, 2, 3, 0, 0, 0, 0, 4,  1, 0, 0, 0, 0, 7,  0, 3, 0,
+     0, 2, 4, 4, 0, 0, 3, 0, 10, 0, 6, 0, 3, 0, 10, 2, 6, 0},
+    {0, 1, 4, 1, 3, 0, 0, 2, 10, 0, 6, 0},
+    {0, 0, 0, 1, 2, 5, 0, 0, 0, 0, 4, 3, 0, 0, 3, 0, 7, 0,
+     2, 4, 0, 2, 0, 0, 8, 0, 6, 0, 1, 0, 8, 2, 6, 0, 1, 0},
+    {0, 0, 0, 0, 7, 0, 2, 4, 6, 0, 1, 0}};
+  std::vector<int32_t> input1_data{-1, 2, 1, 0, 4, -5, 1, 3, 7, -1, 7, 1};
+  std::vector<int32_t> input2_data{4, 1, -3, -1, 1, 6};
```
Comment (binarman) [inline compiler/luci-interpreter/src/kernels/Sub.test.cpp:178]:
```suggestion
  std::vector<dtype> input1_data{-1, 2, 1, 0, 4, -5, 1, 3, 7, -1, 7, 1};
  std::vector<dtype> input2_data{4, 1, -3, -1, 1, 6};
```

File: compiler/luci-interpreter/src/kernels/Sub.test.cpp [hunk 5]
```diff
@@ -162,6 +162,52 @@ TEST_F(SubTest, Float)
   }
 }
 
+template <loco::DataType DType> void CheckInteger(luci_interpreter::IMemoryManager *memory_manager)
+{
+  using dtype = typename loco::DataTypeImpl<DType>::Type;
+  Shape base_shape = {2, 3, 1, 2};
+  std::vector<Shape> test_shapes{{1, 1, 3, 2}, {1, 3, 1, 2}, {2, 1, 3, 1}, {2, 3, 1, 1}};
+  std::vector<std::vector<int32_t>> test_outputs = {
```
Comment (binarman) [inline compiler/luci-interpreter/src/kernels/Sub.test.cpp:170]:
```suggestion
  std::vector<std::vector<dtype>> test_outputs = {
```

File: compiler/luci-interpreter/src/kernels/Sub.test.cpp [hunk 6]
```diff
@@ -162,6 +162,52 @@ TEST_F(SubTest, Float)
...
+
+    SubParams params{};
+    params.activation = Activation::RELU;
+
+    Sub kernel(&input1_tensor, &input2_tensor, &output_tensor, params);
+    kernel.configure();
+    memory_manager->allocate_memory(output_tensor);
+    kernel.execute();
+
+    EXPECT_THAT(extractTensorData<int32_t>(output_tensor), test_outputs[i])
```
Comment (binarman) [inline compiler/luci-interpreter/src/kernels/Sub.test.cpp:194]:
```suggestion
    EXPECT_THAT(extractTensorData<dtype>(output_tensor), test_outputs[i])
```

File: compiler/luci-interpreter/src/kernels/Sub.test.cpp (hunk 1 above)
Comment (BalyshevArtem) [reply compiler/luci-interpreter/src/kernels/Sub.test.cpp]: fix

**Review by binarman — APPROVED** [review]
LGTM

File: compiler/luci-interpreter/src/kernels/Sub.cpp [hunk 7]
```diff
@@ -82,6 +89,44 @@ void Sub::evalFloat() const
...
+  }
+  else
+  {
+    assert((std::is_same<T, int64_t>::value));
+    int64_t activation_min{};
+    int64_t activation_max{};
+    calculateActivationRange(_params.activation, &activation_min, &activation_max);
+    params.int64_activation_min = activation_min;
+    params.int64_activation_max = activation_max;
+  }
```
Comment (m-bronnikov) [inline compiler/luci-interpreter/src/kernels/Sub.cpp:111]: (optional) May be simplify a bit?

```suggestion
calculateActivationRange(_params.activation, &params.quantized_activation_min, &params.quantized_activation_max);
calculateActivationRange(_params.activation, &params.int64_activation_min, &params.int64_activation_max);
```

**Review by m-bronnikov — APPROVED** [review]
LGTM

File: compiler/luci-interpreter/src/kernels/Sub.cpp (hunk 7 above)
Comment (binarman) [reply compiler/luci-interpreter/src/kernels/Sub.cpp:111]: let's refactor this in separate PR, I think we can find same code in other operations

**Pull request merged**